commit 3f8a12f9cf6d9f8091e05b61952409ed9623257d
Author:     Neil Schemenauer <nas-github@arctrix.com>
AuthorDate: Thu Jan 8 01:04:44 2026 -0800
Commit:     GitHub <noreply@github.com>
CommitDate: Thu Jan 8 10:04:44 2026 +0100

    fix(tests): add events_checker() fixture(#1151)
    
    * Wait for directory reader thread to start.
    
    * fix: add events_checker() fixture, use for tests
    
    This makes the unit tests more reliable by allowing events to be
    received in different orders and by allowing additional (non-expected)
    events.
    
    * Check for DirModifiedEvent on file create.
    
    This matches the original test logic, before addition of the
    events_checker() fixture.
    
    * Make events_checker() validate order by default.
    
    Rather than allow events in any order, match the order of the events as
    added.  On MacOS, fsevents does not produce events in a determistic
    order so allow any ordering on that platform.
    
    * Use events_checker() as a context manager.
    
    This reduces the code a little.  Use `ec` as the short local name of
    the checker instance.
    
    Add `verbose` argument to `events_checker()` that turns on verbose
    debugging output.
    
    * Fix lint warnings.
    
    * Unit test fixes for MacOS.
    
    Remove events that are sometimes not generated.
    
    * Fix unreliable test for Window.
    
    * Fix lint warning, improve comment.

diff --git a/src/watchdog/observers/winapi.py b/src/watchdog/observers/winapi.py
index 0e6a131..6bdd5a1 100644
--- a/src/watchdog/observers/winapi.py
+++ b/src/watchdog/observers/winapi.py
@@ -412,6 +412,7 @@ class DirectoryChangeReader:
         # missed, we re-use these for each call to _run_inner().
         event_buffer = ctypes.create_string_buffer(BUFFER_SIZE)
         nbytes = DWORD()
+        self._buf_queue.put(b'')  # indicates that this thread has started
         try:
             while not self._should_stop:
                 self._run_inner(handle, event_buffer, nbytes)
@@ -430,6 +431,11 @@ class DirectoryChangeReader:
                 return  # stop already called, do not start
             self._reader_thread = threading.Thread(target=self._run)
         self._reader_thread.start()
+        # Wait for empty bytes object, indicating reader thread has started.
+        # This reduces the time window between this method returning and the
+        # actual ReadDirectoryChangesW() call, which reduces the time that
+        # events could be missed.
+        _ = self._buf_queue.get(timeout=2)
 
     def stop(self) -> None:
         with self._lock:
diff --git a/tests/conftest.py b/tests/conftest.py
index b0f4ce5..4411b53 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -8,7 +8,7 @@ from functools import partial
 
 import pytest
 
-from .utils import ExpectEvent, Helper, P, StartWatching, TestEventQueue
+from .utils import EventsChecker, ExpectEvent, Helper, P, StartWatching, TestEventQueue
 
 
 @pytest.fixture
@@ -78,3 +78,8 @@ def start_watching_fixture(helper: Helper) -> StartWatching:
 @pytest.fixture(name="expect_event")
 def expect_event_fixture(helper: Helper) -> ExpectEvent:
     return helper.expect_event
+
+
+@pytest.fixture(name="events_checker")
+def events_checker_fixture(helper: Helper) -> EventsChecker:
+    return helper.events_checker
diff --git a/tests/test_emitter.py b/tests/test_emitter.py
index 5d0aa28..3b30b46 100644
--- a/tests/test_emitter.py
+++ b/tests/test_emitter.py
@@ -27,7 +27,7 @@ from watchdog.utils import platform
 from .shell import mkdir, mkfile, mv, rm, touch
 
 if TYPE_CHECKING:
-    from .utils import ExpectEvent, P, StartWatching, TestEventQueue
+    from .utils import EventsChecker, P, StartWatching, TestEventQueue
 
 logging.basicConfig(level=logging.DEBUG)
 logger = logging.getLogger(__name__)
@@ -39,113 +39,106 @@ if platform.is_darwin():
     fsevents_logger.setLevel(logging.DEBUG)
 
 
-def rerun_filter(exc, *args):
-    time.sleep(5)
-    return bool(issubclass(exc[0], Empty) and platform.is_windows())
-
-
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_create(p: P, event_queue: TestEventQueue, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_create(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     start_watching()
     open(p("a"), "a").close()
 
-    expect_event(FileCreatedEvent(p("a")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
-
-    if platform.is_linux():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("a")
-        assert isinstance(event, FileOpenedEvent)
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("a")
-        assert isinstance(event, FileClosedEvent)
+    with events_checker() as ec:
+        if platform.is_darwin():
+            ec.add(FileCreatedEvent, "a")
+        else:
+            ec.add(FileCreatedEvent, "a")
+            if not platform.is_windows():
+                ec.add(DirModifiedEvent, ".")
+            if platform.is_linux():
+                ec.add(FileOpenedEvent, "a")
+                ec.add(FileClosedEvent, "a")
+            if not platform.is_windows():
+                ec.add(DirModifiedEvent, ".")
 
 
 @pytest.mark.skipif(not platform.is_linux(), reason="FileClosed*Event only supported in GNU/Linux")
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_closed(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
+def test_closed(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     with open(p("a"), "a"):
         start_watching()
 
-    # After file creation/open in append mode
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("a")
-    assert isinstance(event, FileClosedEvent)
+    with events_checker() as ec:
+        # After file creation/open in append mode
+        ec.add(FileClosedEvent, "a")
 
-    event = event_queue.get(timeout=5)[0]
-    assert os.path.normpath(event.src_path) == os.path.normpath(p(""))
-    assert isinstance(event, DirModifiedEvent)
+        ec.add(DirModifiedEvent, ".")
 
     # After read-only, only IN_CLOSE_NOWRITE is emitted
     open(p("a")).close()
 
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("a")
-    assert isinstance(event, FileOpenedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("a")
-    assert isinstance(event, FileClosedNoWriteEvent)
-
-    assert event_queue.empty()
+    with events_checker() as ec:
+        ec.add(FileOpenedEvent, "a")
+        ec.add(FileClosedNoWriteEvent, "a")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 @pytest.mark.skipif(
     platform.is_darwin() or platform.is_windows(),
     reason="Windows and macOS enforce proper encoding",
 )
-def test_create_wrong_encoding(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
+def test_create_wrong_encoding(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     start_watching()
-    open(p("a_\udce4"), "a").close()
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("a_\udce4")
-    assert isinstance(event, FileCreatedEvent)
+    filename = "a_\udce4"
+    open(p(filename), "a").close()
 
-    if not platform.is_windows():
-        event = event_queue.get(timeout=5)[0]
-        assert os.path.normpath(event.src_path) == os.path.normpath(p(""))
-        assert isinstance(event, DirModifiedEvent)
+    with events_checker() as ec:
+        ec.add(FileCreatedEvent, filename)
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, ".")
+            ec.add(FileOpenedEvent, filename)
+            ec.add(FileClosedEvent, filename)
+            ec.add(DirModifiedEvent, ".")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_delete(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_delete(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkfile(p("a"))
 
     start_watching()
     rm(p("a"))
 
-    expect_event(FileDeletedEvent(p("a")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
+    with events_checker() as ec:
+        if platform.is_darwin():
+            ec.add(DirModifiedEvent, ".")
+        else:
+            ec.add(FileDeletedEvent, "a")
+            if not platform.is_windows():
+                ec.add(DirModifiedEvent, ".")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_modify(p: P, event_queue: TestEventQueue, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_modify(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkfile(p("a"))
     start_watching()
 
     touch(p("a"))
 
-    if platform.is_linux():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("a")
-        assert isinstance(event, FileOpenedEvent)
+    with events_checker() as ec:
+        if platform.is_darwin():
+            ec.add(FileModifiedEvent, "a")
+        else:
+            if platform.is_linux():
+                ec.add(FileOpenedEvent, "a")
+            ec.add(FileModifiedEvent, "a")
+            if platform.is_linux():
+                ec.add(FileClosedEvent, "a")
+            if not platform.is_windows():
+                ec.add(DirModifiedEvent, ".")
 
-    expect_event(FileModifiedEvent(p("a")))
 
-    if platform.is_linux():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("a")
-        assert isinstance(event, FileClosedEvent)
-
-
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_chmod(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_chmod(p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker) -> None:
     mkfile(p("a"))
     start_watching()
 
@@ -153,14 +146,16 @@ def test_chmod(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -
     # allows setting the read-only flag.
     os.chmod(p("a"), stat.S_IREAD)
 
-    expect_event(FileModifiedEvent(p("a")))
+    with events_checker() as ec:
+        ec.add(FileModifiedEvent, "a")
 
     # Reset permissions to allow cleanup.
     os.chmod(p("a"), stat.S_IWRITE)
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_move(p: P, event_queue: TestEventQueue, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_move_simple(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     mkdir(p("dir2"))
     mkfile(p("dir1", "a"))
@@ -168,32 +163,22 @@ def test_move(p: P, event_queue: TestEventQueue, start_watching: StartWatching,
 
     mv(p("dir1", "a"), p("dir2", "b"))
 
-    if not platform.is_windows():
-        expect_event(FileMovedEvent(p("dir1", "a"), p("dir2", "b")))
-    else:
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("dir1", "a")
-        assert isinstance(event, FileDeletedEvent)
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("dir2", "b")
-        assert isinstance(event, FileCreatedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path in [p("dir1"), p("dir2")]
-    assert isinstance(event, DirModifiedEvent)
-
-    if not platform.is_windows():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path in [p("dir1"), p("dir2")]
-        assert isinstance(event, DirModifiedEvent)
+    with events_checker() as ec:
+        if not platform.is_windows():
+            ec.add(FileMovedEvent, "dir1/a", dest_path="dir2/b")
+            ec.add(DirModifiedEvent, "dir1")
+            ec.add(DirModifiedEvent, "dir2")
+        else:
+            ec.add(FileDeletedEvent, "dir1/a")
+            ec.add(FileCreatedEvent, "dir2/b")
+            ec.add(DirModifiedEvent, "dir2")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 def test_case_change(
     p: P,
     event_queue: TestEventQueue,
     start_watching: StartWatching,
-    expect_event: ExpectEvent,
+    events_checker: EventsChecker,
 ) -> None:
     mkdir(p("dir1"))
     mkdir(p("dir2"))
@@ -202,28 +187,20 @@ def test_case_change(
 
     mv(p("dir1", "file"), p("dir2", "FILE"))
 
-    if not platform.is_windows():
-        expect_event(FileMovedEvent(p("dir1", "file"), p("dir2", "FILE")))
-    else:
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("dir1", "file")
-        assert isinstance(event, FileDeletedEvent)
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("dir2", "FILE")
-        assert isinstance(event, FileCreatedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path in [p("dir1"), p("dir2")]
-    assert isinstance(event, DirModifiedEvent)
-
-    if not platform.is_windows():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path in [p("dir1"), p("dir2")]
-        assert isinstance(event, DirModifiedEvent)
+    with events_checker() as ec:
+        if not platform.is_windows():
+            ec.add(FileMovedEvent, "dir1/file", dest_path="dir2/FILE")
+            ec.add(DirModifiedEvent, "dir1")
+            ec.add(DirModifiedEvent, "dir2")
+        else:
+            ec.add(FileDeletedEvent, "dir1/file")
+            ec.add(FileCreatedEvent, "dir2/FILE")
+            ec.add(DirModifiedEvent, "dir2")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_move_to(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_move_to(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     mkdir(p("dir2"))
     mkfile(p("dir1", "a"))
@@ -231,28 +208,32 @@ def test_move_to(p: P, start_watching: StartWatching, expect_event: ExpectEvent)
 
     mv(p("dir1", "a"), p("dir2", "b"))
 
-    expect_event(FileCreatedEvent(p("dir2", "b")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p("dir2")))
+    with events_checker() as ec:
+        ec.add(FileCreatedEvent, "dir2/b")
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, "dir2")
 
 
 @pytest.mark.skipif(not platform.is_linux(), reason="InotifyFullEmitter only supported in Linux")
-def test_move_to_full(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
+def test_move_to_full(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     mkdir(p("dir2"))
     mkfile(p("dir1", "a"))
     start_watching(path=p("dir2"), use_full_emitter=True)
     mv(p("dir1", "a"), p("dir2", "b"))
 
-    event = event_queue.get(timeout=5)[0]
-    assert isinstance(event, FileMovedEvent)
-    assert event.dest_path == p("dir2", "b")
-    assert event.src_path == ""  # Should be blank since the path was not watched
+    with events_checker() as ec:
+        # The src_path should be blank since the path was not watched
+        ec.add(FileMovedEvent, "", dest_path="dir2/b")
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, "dir2")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_move_from(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_move_from(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     mkdir(p("dir2"))
     mkfile(p("dir1", "a"))
@@ -260,28 +241,32 @@ def test_move_from(p: P, start_watching: StartWatching, expect_event: ExpectEven
 
     mv(p("dir1", "a"), p("dir2", "b"))
 
-    expect_event(FileDeletedEvent(p("dir1", "a")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p("dir1")))
+    with events_checker() as ec:
+        ec.add(FileDeletedEvent, "dir1/a")
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, "dir1")
 
 
 @pytest.mark.skipif(not platform.is_linux(), reason="InotifyFullEmitter only supported in Linux")
-def test_move_from_full(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
+def test_move_from_full(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     mkdir(p("dir2"))
     mkfile(p("dir1", "a"))
     start_watching(path=p("dir1"), use_full_emitter=True)
     mv(p("dir1", "a"), p("dir2", "b"))
 
-    event = event_queue.get(timeout=5)[0]
-    assert isinstance(event, FileMovedEvent)
-    assert event.src_path == p("dir1", "a")
-    assert event.dest_path == ""  # Should be blank since path not watched
+    with events_checker() as ec:
+        # dest_path should be blank since not watched
+        ec.add(FileMovedEvent, "dir1/a", dest_path="")
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, "dir1")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_separate_consecutive_moves(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_separate_consecutive_moves(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     mkfile(p("dir1", "a"))
     mkfile(p("b"))
@@ -289,31 +274,28 @@ def test_separate_consecutive_moves(p: P, start_watching: StartWatching, expect_
     mv(p("dir1", "a"), p("c"))
     mv(p("b"), p("dir1", "d"))
 
-    dir_modif = DirModifiedEvent(p("dir1"))
-    a_deleted = FileDeletedEvent(p("dir1", "a"))
-    d_created = FileCreatedEvent(p("dir1", "d"))
+    with events_checker() as ec:
+        if not platform.is_windows():
+            ec.add(FileDeletedEvent, "dir1/a")
+            ec.add(DirModifiedEvent, "dir1")
+            ec.add(FileCreatedEvent, "dir1/d")
+            ec.add(DirModifiedEvent, "dir1")
+        else:
+            ec.add(FileDeletedEvent, "dir1/a")
+            ec.add(FileCreatedEvent, "dir1/d")
 
-    expected_events = [a_deleted, dir_modif, d_created, dir_modif]
 
-    if platform.is_windows():
-        expected_events = [a_deleted, d_created]
-
-    if platform.is_bsd():
-        # Due to the way kqueue works, we can't really order
-        # 'Created' and 'Deleted' events in time, so creation queues first
-        expected_events = [d_created, a_deleted, dir_modif, dir_modif]
-
-    for expected_event in expected_events:
-        expect_event(expected_event)
-
-
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 @pytest.mark.skipif(platform.is_bsd(), reason="BSD create another set of events for this test")
-def test_delete_self(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_delete_self(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1"))
     emitter = start_watching(path=p("dir1"))
     rm(p("dir1"), recursive=True)
-    expect_event(DirDeletedEvent(p("dir1")))
+
+    with events_checker() as ec:
+        ec.add(DirDeletedEvent, "dir1")
+
     emitter.join(5)
     assert not emitter.is_alive()
 
@@ -322,7 +304,9 @@ def test_delete_self(p: P, start_watching: StartWatching, expect_event: ExpectEv
     platform.is_windows() or platform.is_bsd(),
     reason="Windows|BSD create another set of events for this test",
 )
-def test_fast_subdirectory_creation_deletion(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
+def test_fast_subdirectory_creation_deletion(
+    p: P, events_checker: EventsChecker, event_queue: TestEventQueue, start_watching: StartWatching
+) -> None:
     root_dir = p("dir1")
     sub_dir = p("dir1", "subdir1")
     times = 30
@@ -332,28 +316,15 @@ def test_fast_subdirectory_creation_deletion(p: P, event_queue: TestEventQueue,
         mkdir(sub_dir)
         rm(sub_dir, recursive=True)
         time.sleep(0.1)  # required for macOS emitter to catch up with us
-    count = {DirCreatedEvent: 0, DirModifiedEvent: 0, DirDeletedEvent: 0}
-    etype_for_dir = {
-        DirCreatedEvent: sub_dir,
-        DirModifiedEvent: root_dir,
-        DirDeletedEvent: sub_dir,
-    }
-    for _ in range(times * 4):
-        event = event_queue.get(timeout=5)[0]
-        logger.debug(event)
-        etype = type(event)
-        count[etype] += 1
-        assert event.src_path == etype_for_dir[etype]
-        assert count[DirCreatedEvent] >= count[DirDeletedEvent]
-        assert count[DirCreatedEvent] + count[DirDeletedEvent] >= count[DirModifiedEvent]
-    assert count == {
-        DirCreatedEvent: times,
-        DirModifiedEvent: times * 2,
-        DirDeletedEvent: times,
-    }
-
-
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
+
+    with events_checker() as ec:
+        for _ in range(times):
+            ec.add(DirCreatedEvent, "dir1/subdir1")
+            ec.add(DirModifiedEvent, "dir1")
+            ec.add(DirDeletedEvent, "dir1/subdir1")
+            ec.add(DirModifiedEvent, "dir1")
+
+
 def test_passing_unicode_should_give_unicode(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
     start_watching(path=str(p()))
     mkfile(p("a"))
@@ -363,7 +334,7 @@ def test_passing_unicode_should_give_unicode(p: P, event_queue: TestEventQueue,
 
 @pytest.mark.skipif(
     platform.is_windows(),
-    reason="Windows ReadDirectoryChangesW supports only" " unicode for paths.",
+    reason="Windows ReadDirectoryChangesW supports only unicode for paths.",
 )
 def test_passing_bytes_should_give_bytes(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
     start_watching(path=p().encode())
@@ -372,126 +343,137 @@ def test_passing_bytes_should_give_bytes(p: P, event_queue: TestEventQueue, star
     assert isinstance(event.src_path, bytes)
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
-def test_recursive_on(p: P, event_queue: TestEventQueue, start_watching: StartWatching) -> None:
+def test_recursive_on(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     mkdir(p("dir1", "dir2", "dir3"), parents=True)
     start_watching()
     touch(p("dir1", "dir2", "dir3", "a"))
 
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("dir1", "dir2", "dir3", "a")
-    assert isinstance(event, FileCreatedEvent)
+    with events_checker() as ec:
+        if platform.is_darwin():
+            ec.add(FileCreatedEvent, "dir1/dir2/dir3/a")
+            ec.add(DirModifiedEvent, "dir1/dir2/dir3")
+            ec.add(FileModifiedEvent, "dir1/dir2/dir3/a")
+        elif platform.is_windows():
+            ec.add(FileCreatedEvent, "dir1/dir2/dir3/a")
+            ec.add(FileModifiedEvent, "dir1/dir2/dir3/a")
+        else:
+            ec.add(FileCreatedEvent, "dir1/dir2/dir3/a")
+            ec.add(DirModifiedEvent, "dir1/dir2/dir3")
 
-    if not platform.is_windows():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("dir1", "dir2", "dir3")
-        assert isinstance(event, DirModifiedEvent)
+            if platform.is_linux():
+                ec.add(FileOpenedEvent, "dir1/dir2/dir3/a")
+
+            if not platform.is_bsd():
+                ec.add(FileModifiedEvent, "dir1/dir2/dir3/a")
+
+            if platform.is_linux():
+                ec.add(FileClosedEvent, "dir1/dir2/dir3/a")
+
+            ec.add(DirModifiedEvent, "dir1/dir2/dir3")
 
-        if platform.is_linux():
-            event = event_queue.get(timeout=5)[0]
-            assert event.src_path == p("dir1", "dir2", "dir3", "a")
-            assert isinstance(event, FileOpenedEvent)
 
-        if not platform.is_bsd():
-            event = event_queue.get(timeout=5)[0]
-            assert event.src_path == p("dir1", "dir2", "dir3", "a")
-            assert isinstance(event, FileModifiedEvent)
+def check_empty_queue(event_queue: TestEventQueue) -> None:
+    events = []
+    while True:
+        try:
+            _ = event_queue.get(timeout=4)
+        except Empty:
+            break
+    assert not events, "queue was not empty"
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 def test_recursive_off(
     p: P,
     event_queue: TestEventQueue,
     start_watching: StartWatching,
-    expect_event: ExpectEvent,
+    events_checker: EventsChecker,
 ) -> None:
     mkdir(p("dir1"))
     start_watching(recursive=False)
     touch(p("dir1", "a"))
 
-    with pytest.raises(Empty):
-        event_queue.get(timeout=5)
+    check_empty_queue(event_queue)
 
     mkfile(p("b"))
-    expect_event(FileCreatedEvent(p("b")))
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
 
-        if platform.is_linux():
-            expect_event(FileOpenedEvent(p("b")))
-            expect_event(FileClosedEvent(p("b")))
+    with events_checker() as ec:
+        if platform.is_darwin():
+            ec.add(FileCreatedEvent, "b")
+            ec.add(DirModifiedEvent, ".")
+        else:
+            ec.add(FileCreatedEvent, "b")
+            if not platform.is_windows():
+                ec.add(DirModifiedEvent, ".")
+                if platform.is_linux():
+                    ec.add(FileOpenedEvent, "b")
+                    ec.add(FileClosedEvent, "b")
+                ec.add(DirModifiedEvent, ".")
 
     # currently limiting these additional events to macOS only, see https://github.com/gorakhargosh/watchdog/pull/779
     if platform.is_darwin():
         mkdir(p("dir1", "dir2"))
-        with pytest.raises(Empty):
-            event_queue.get(timeout=5)
+        check_empty_queue(event_queue)
         mkfile(p("dir1", "dir2", "somefile"))
-        with pytest.raises(Empty):
-            event_queue.get(timeout=5)
+        check_empty_queue(event_queue)
 
         mkdir(p("dir3"))
-        expect_event(DirModifiedEvent(p()))  # the contents of the parent directory changed
+        with events_checker() as ec:
+            ec.add(DirModifiedEvent, ".")  # the contents of the parent directory changed
 
         mv(p("dir1", "dir2", "somefile"), p("somefile"))
-        expect_event(FileMovedEvent(p("dir1", "dir2", "somefile"), p("somefile")))
-        expect_event(DirModifiedEvent(p()))
+        with events_checker() as ec:
+            ec.add(FileMovedEvent, "dir1/dir2/somefile", dest_path="somefile")
+            ec.add(DirModifiedEvent, ".")
 
         mv(p("dir1", "dir2"), p("dir2"))
-        expect_event(DirMovedEvent(p("dir1", "dir2"), p("dir2")))
-        expect_event(DirModifiedEvent(p()))
+        with events_checker() as ec:
+            ec.add(DirMovedEvent, "dir1/dir2", dest_path="dir2")
+            ec.add(DirModifiedEvent, ".")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 def test_renaming_top_level_directory(
     p: P,
     event_queue: TestEventQueue,
     start_watching: StartWatching,
-    expect_event: ExpectEvent,
+    events_checker: EventsChecker,
 ) -> None:
     start_watching()
 
     mkdir(p("a"))
-    expect_event(DirCreatedEvent(p("a")))
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
+    with events_checker() as ec:
+        ec.add(DirCreatedEvent, "a")
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, ".")
 
     mkdir(p("a", "b"))
-    expect_event(DirCreatedEvent(p("a", "b")))
-    expect_event(DirModifiedEvent(p("a")))
+    with events_checker() as ec:
+        ec.add(DirCreatedEvent, "a/b")
+        if not platform.is_windows():
+            ec.add(DirModifiedEvent, "a")
 
     mv(p("a"), p("a2"))
-    expect_event(DirMovedEvent(p("a"), p("a2")))
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
-        expect_event(DirModifiedEvent(p()))
-    expect_event(DirMovedEvent(p("a", "b"), p("a2", "b"), is_synthetic=True))
-
-    if platform.is_bsd():
-        expect_event(DirModifiedEvent(p()))
+    with events_checker() as ec:
+        if platform.is_windows():
+            ec.add(DirMovedEvent, "a", dest_path="a2")
+            ec.add(DirMovedEvent, "a/b", dest_path="a2/b")
+            ec.add(DirModifiedEvent, "a2")
+        else:
+            ec.add(DirMovedEvent, "a", dest_path="a2")
+            ec.add(DirModifiedEvent, ".")
+            ec.add(DirModifiedEvent, ".")
+            ec.add(DirMovedEvent, "a/b", dest_path="a2/b")
 
     open(p("a2", "b", "c"), "a").close()
 
-    # DirModifiedEvent may emitted, but sometimes after waiting time is out.
-    events = []
-    while True:
-        events.append(event_queue.get(timeout=5)[0])
-        if event_queue.empty():
-            break
-
-    assert all(
-        isinstance(e, (FileCreatedEvent, FileMovedEvent, FileOpenedEvent, DirModifiedEvent, FileClosedEvent))
-        for e in events
-    )
-
-    for event in events:
-        if isinstance(event, FileCreatedEvent):
-            assert event.src_path == p("a2", "b", "c")
-        elif isinstance(event, FileMovedEvent):
-            assert event.dest_path == p("a2", "b", "c")
-            assert event.src_path == p("a", "b", "c")
-        elif isinstance(event, DirModifiedEvent):
-            assert event.src_path == p("a2", "b")
+    with events_checker() as ec:
+        ec.add(FileCreatedEvent, "a2/b/c")
+        if platform.is_linux():
+            ec.add(DirModifiedEvent, "a2/b")
+            ec.add(FileOpenedEvent, "a2/b/c")
+            ec.add(FileClosedEvent, "a2/b/c")
+            ec.add(DirModifiedEvent, "a2/b")
 
 
 @pytest.mark.skipif(platform.is_windows(), reason="Windows create another set of events for this test")
@@ -499,42 +481,31 @@ def test_move_nested_subdirectories(
     p: P,
     event_queue: TestEventQueue,
     start_watching: StartWatching,
-    expect_event: ExpectEvent,
+    events_checker: EventsChecker,
 ) -> None:
     mkdir(p("dir1/dir2/dir3"), parents=True)
     mkfile(p("dir1/dir2/dir3", "a"))
     start_watching()
     mv(p("dir1/dir2"), p("dir2"))
 
-    expect_event(DirMovedEvent(p("dir1", "dir2"), p("dir2")))
-    expect_event(DirModifiedEvent(p("dir1")))
-    expect_event(DirModifiedEvent(p()))
-
-    expect_event(DirMovedEvent(p("dir1", "dir2", "dir3"), p("dir2", "dir3"), is_synthetic=True))
-    expect_event(FileMovedEvent(p("dir1", "dir2", "dir3", "a"), p("dir2", "dir3", "a"), is_synthetic=True))
-
-    if platform.is_bsd():
-        event = event_queue.get(timeout=5)[0]
-        assert p(event.src_path) == p()
-        assert isinstance(event, DirModifiedEvent)
-
-        event = event_queue.get(timeout=5)[0]
-        assert p(event.src_path) == p("dir1")
-        assert isinstance(event, DirModifiedEvent)
+    with events_checker() as ec:
+        ec.add(DirMovedEvent, "dir1/dir2", dest_path="dir2")
+        ec.add(DirModifiedEvent, "dir1")
+        ec.add(DirModifiedEvent, ".")
+        ec.add(DirMovedEvent, "dir1/dir2/dir3", dest_path="dir2/dir3")
+        ec.add(FileMovedEvent, "dir1/dir2/dir3/a", dest_path="dir2/dir3/a")
 
     touch(p("dir2/dir3", "a"))
 
-    if platform.is_linux():
-        event = event_queue.get(timeout=5)[0]
-        assert event.src_path == p("dir2/dir3", "a")
-        assert isinstance(event, FileOpenedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("dir2/dir3", "a")
-    assert isinstance(event, FileModifiedEvent)
+    with events_checker() as ec:
+        if platform.is_linux():
+            ec.add(FileOpenedEvent, "dir2/dir3/a")
+        ec.add(FileModifiedEvent, "dir2/dir3/a")
+        if platform.is_linux():
+            ec.add(FileClosedEvent, "dir2/dir3/a")
+            ec.add(DirModifiedEvent, "dir2/dir3")
 
 
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 @pytest.mark.skipif(
     not platform.is_windows(),
     reason="Non-Windows create another set of events for this test",
@@ -543,48 +514,36 @@ def test_move_nested_subdirectories_on_windows(
     p: P,
     event_queue: TestEventQueue,
     start_watching: StartWatching,
+    events_checker: EventsChecker,
 ) -> None:
     mkdir(p("dir1/dir2/dir3"), parents=True)
     mkfile(p("dir1/dir2/dir3", "a"))
     start_watching(path=p(""))
     mv(p("dir1/dir2"), p("dir2"))
 
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("dir1", "dir2")
-    assert isinstance(event, FileDeletedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("dir2")
-    assert isinstance(event, DirCreatedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("dir2", "dir3")
-    assert isinstance(event, DirCreatedEvent)
-
-    event = event_queue.get(timeout=5)[0]
-    assert event.src_path == p("dir2", "dir3", "a")
-    assert isinstance(event, FileCreatedEvent)
+    with events_checker() as ec:
+        ec.add(FileDeletedEvent, "dir1/dir2")
+        ec.add(DirCreatedEvent, "dir2")
+        ec.add(DirCreatedEvent, "dir2/dir3")
+        ec.add(FileCreatedEvent, "dir2/dir3/a")
+        ec.add(DirModifiedEvent, "dir2")
+        if False:
+            # The following event is not consistently generated.
+            ec.add(DirModifiedEvent, "dir2/dir3")
+        else:
+            # Allow test to pass if above event is present or not
+            ec.allow_extra_events()
 
     touch(p("dir2/dir3", "a"))
 
-    events = []
-    while True:
-        events.append(event_queue.get(timeout=5)[0])
-        if event_queue.empty():
-            break
-
-    assert all(isinstance(e, (FileModifiedEvent, DirModifiedEvent)) for e in events)
+    with events_checker() as ec:
+        ec.add(FileModifiedEvent, "dir2/dir3/a")
 
-    for event in events:
-        if isinstance(event, FileModifiedEvent):
-            assert event.src_path == p("dir2", "dir3", "a")
-        elif isinstance(event, DirModifiedEvent):
-            assert event.src_path in [p("dir2"), p("dir2", "dir3")]
 
-
-@pytest.mark.flaky(max_runs=5, min_passes=1, rerun_filter=rerun_filter)
 @pytest.mark.skipif(platform.is_bsd(), reason="BSD create another set of events for this test")
-def test_file_lifecyle(p: P, start_watching: StartWatching, expect_event: ExpectEvent) -> None:
+def test_file_lifecyle(
+    p: P, event_queue: TestEventQueue, start_watching: StartWatching, events_checker: EventsChecker
+) -> None:
     start_watching()
 
     mkfile(p("a"))
@@ -592,30 +551,28 @@ def test_file_lifecyle(p: P, start_watching: StartWatching, expect_event: Expect
     mv(p("a"), p("b"))
     rm(p("b"))
 
-    expect_event(FileCreatedEvent(p("a")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
-
-    if platform.is_linux():
-        expect_event(FileOpenedEvent(p("a")))
-        expect_event(FileClosedEvent(p("a")))
-        expect_event(DirModifiedEvent(p()))
-        expect_event(FileOpenedEvent(p("a")))
-
-    expect_event(FileModifiedEvent(p("a")))
-
-    if platform.is_linux():
-        expect_event(FileClosedEvent(p("a")))
-        expect_event(DirModifiedEvent(p()))
-
-    expect_event(FileMovedEvent(p("a"), p("b")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
-        expect_event(DirModifiedEvent(p()))
-
-    expect_event(FileDeletedEvent(p("b")))
-
-    if not platform.is_windows():
-        expect_event(DirModifiedEvent(p()))
+    with events_checker() as ec:
+        if platform.is_linux():
+            ec.add(FileCreatedEvent, "a")
+            ec.add(DirModifiedEvent, ".")
+
+            ec.add(FileOpenedEvent, "a")
+            ec.add(FileClosedEvent, "a")
+            ec.add(DirModifiedEvent, ".")
+
+            ec.add(FileOpenedEvent, "a")
+            ec.add(FileModifiedEvent, "a")
+            ec.add(FileClosedEvent, "a")
+            ec.add(DirModifiedEvent, ".")
+
+            ec.add(FileMovedEvent, "a", dest_path="b")
+            ec.add(DirModifiedEvent, ".")
+            ec.add(DirModifiedEvent, ".")
+
+            ec.add(FileDeletedEvent, "b")
+            ec.add(DirModifiedEvent, ".")
+        else:
+            ec.add(FileCreatedEvent, "a")
+            ec.add(FileModifiedEvent, "a")
+            ec.add(FileMovedEvent, "a", dest_path="b")
+            ec.add(FileDeletedEvent, "b")
diff --git a/tests/test_fsevents.py b/tests/test_fsevents.py
index 47185eb..7d033dd 100644
--- a/tests/test_fsevents.py
+++ b/tests/test_fsevents.py
@@ -4,6 +4,7 @@ import contextlib
 
 import pytest
 
+from watchdog.events import DirCreatedEvent
 from watchdog.utils import platform
 
 if not platform.is_darwin():
@@ -29,7 +30,7 @@ from watchdog.observers.fsevents import FSEventsEmitter
 from .shell import touch
 
 if TYPE_CHECKING:
-    from .utils import P, StartWatching, TestEventQueue
+    from .utils import EventsChecker, P, StartWatching
 
 logging.basicConfig(level=logging.DEBUG)
 logger = logging.getLogger(__name__)
@@ -228,7 +229,7 @@ def test_unschedule_removed_folder(observer: BaseObserver, p: P) -> None:
     observer.unschedule(w)
 
 
-def test_converting_cfstring_to_pyunicode(p: P, start_watching: StartWatching, event_queue: TestEventQueue) -> None:
+def test_converting_cfstring_to_pyunicode(p: P, start_watching: StartWatching, events_checker: EventsChecker) -> None:
     """See https://github.com/gorakhargosh/watchdog/issues/762"""
 
     tmpdir = p()
@@ -238,8 +239,8 @@ def test_converting_cfstring_to_pyunicode(p: P, start_watching: StartWatching, e
 
     try:
         mkdir(p(dirname))
-        event, _ = event_queue.get()
-        assert event.src_path.endswith(dirname)
+        with events_checker() as ec:
+            ec.add(DirCreatedEvent, dirname)
     finally:
         emitter.stop()
 
diff --git a/tests/test_observers_winapi.py b/tests/test_observers_winapi.py
index 87936d1..42d509c 100644
--- a/tests/test_observers_winapi.py
+++ b/tests/test_observers_winapi.py
@@ -77,6 +77,10 @@ def test___init__(event_queue, emitter):
         except Empty:
             break
         else:
+            if event.event_type == "modified":
+                # On Windows can get one or two modified events.  Ignore
+                # them to make test more deterministic.
+                continue
             got.add(event)
 
     assert expected == got
diff --git a/tests/utils.py b/tests/utils.py
index 759a9b0..948562f 100644
--- a/tests/utils.py
+++ b/tests/utils.py
@@ -4,8 +4,8 @@ import dataclasses
 import os
 import subprocess
 import sys
-from queue import Queue
-from typing import Protocol
+from queue import Empty, Queue
+from typing import Any, Protocol
 
 from watchdog.events import FileSystemEvent
 from watchdog.observers.api import EventEmitter, ObservedWatch
@@ -81,6 +81,16 @@ class Helper:
 
         return emitter
 
+    def events_checker(self, *, verbose: bool = False) -> _EventsChecker:
+        """Utility function to create a new event checker instance.  Use as a context
+        manager and call add() to add events to check for.  The events in the queue
+        will automatically be checked when the "with" block is exited.
+
+        If `verbose` is true, print details of expected and found events to stdout (to assist with
+        developing or debugging tests).
+        """
+        return _EventsChecker(self, verbose=verbose)
+
     def expect_event(self, expected_event: FileSystemEvent, timeout: float = 2) -> None:
         """Utility function to wait up to `timeout` seconds for an `event_type` for `path` to show up in the queue.
 
@@ -125,3 +135,155 @@ def run_isolated_test(path):
         raise
 
     assert p.returncode == 0
+
+
+class EventsChecker(Protocol):
+    def __call__(self) -> _EventsChecker: ...
+
+
+@dataclasses.dataclass()
+class _ExpectedEvent:
+    """Specifies what kind of event we are expecting.  The `expected_class` is required,
+    everything else is optional (not checked if it is None).
+    """
+
+    expected_class: type
+    src_path: str | None = None
+    dest_path: str | None = None
+
+
+class _EventsChecker:
+
+    expected_events: list[_ExpectedEvent]
+
+    def __init__(self, helper: Helper, *, verbose: bool = False):
+        self.tmp = helper.tmp
+        self.event_queue = helper.event_queue
+        self.expected_events = []
+        self._verbose = verbose
+        # If true, check that we receive exactly the expected events in the
+        # specified order.
+        self._validate_order = True
+        # If true, allow extra events in addition to expected ones
+        self._allow_extra = False
+        if platform.is_darwin():
+            # The fsevents API gives back events in a non-specific order.
+            self._validate_order = False
+            # Sometimes there are addtional DirModifiedEvents returned as well.
+            self._allow_extra = True
+
+    def allow_extra_events(self) -> None:
+        """Allow events not specified by add()."""
+        self._allow_extra = True
+
+    def _debug(self, *args: Any) -> None:
+        if self._verbose:
+            print(" == events checker == ", *args)  # noqa: T201
+
+    def _make_path(self, path: str | None) -> str | None:
+        if path is None:
+            return None
+        if path == "":
+            # an empty path is kept as-is
+            return ""
+        # convert to platform specific path and normalize
+        parts = path.split("/")
+        path = os.path.join(self.tmp, *parts)
+        return os.path.normpath(path)
+
+    def add(self, expected_class: type, src_path: str | None = None, dest_path: str | None = None) -> None:
+        """Add details for an expected event.  The `expected_class` argument
+        is required, everything else is optional.  The order that events are
+        received does not matter but adding the same kind of event more than
+        once will require that it appears more than once.
+
+        Note that paths are provided as relative to `tmp` and using the forward
+        slash separator. They will be converted to absolute paths using `tmp`
+        and normalized.  An empty path, i.e. "", is kept as-is.
+        """
+        self.expected_events.append(_ExpectedEvent(expected_class, src_path, dest_path))
+
+    def _match_event(self, event: FileSystemEvent, expected_event: _ExpectedEvent) -> bool:
+        if not isinstance(event, expected_event.expected_class):
+            return False  # wrong class
+        src_path = self._make_path(expected_event.src_path)
+        if src_path is not None and src_path != event.src_path:
+            return False  # wrong src_path
+        dest_path = self._make_path(expected_event.dest_path)
+        if dest_path is not None and dest_path != event.dest_path:
+            return False  # wrong dest_path
+        self._debug("matched:", expected_event)
+        return True
+
+    def _check_events_without_order( self, found_events: list[FileSystemEvent]) -> None:
+        # Check that events received contain the expected events.  This is
+        # an inefficient way to do it, O(n*m) but since the list of expected
+        # events should be fairly short, this is okay.  Using a list allows
+        # the expected events to contain the same kind of event more than
+        # once.
+        expected_events = list(self.expected_events)
+        matched_events = []
+        for event in found_events:
+            for i, expected_event in enumerate(expected_events):
+                if self._match_event(event, expected_event):
+                    del expected_events[i]
+                    matched_events.append(event)
+                    break
+        if expected_events:
+            # Fail, we did not find some of the expected events.
+            if self._verbose:
+                self._debug("missing:")
+                for e in expected_events:
+                    self._debug("  ", e.expected_class.__name__, e.src_path)
+            assert not expected_events, "some expected events not found"
+        if not self._allow_extra:  # noqa: SIM102
+            # Check that we did not receive extra events
+            if len(matched_events) < len(found_events):
+                for event in found_events:
+                    if event not in matched_events:
+                        self._debug("unexpected", event)
+                msg = "received unexpected events"
+                raise AssertionError(msg)
+
+    def _check_events_with_order( self, found_events: list[FileSystemEvent]) -> None:
+        if not self._allow_extra:
+            # we need exactly the number of expected events
+            assert len(found_events) == len(self.expected_events), "wrong number of events"
+        for event, expected_event in zip(found_events, self.expected_events):
+            assert self._match_event(event, expected_event), "did not find expected event"
+
+    def check_events(self, timeout: float = 2) -> None:
+        """Read events from the events queue (waiting for up to `timeout` for
+        new events to appear).  Confirm that expected events, as specified by
+        calling add(), appear in the sequence of events receieved.
+        """
+        if self._verbose:
+            self._debug("expecting:")
+            for e in self.expected_events:
+                self._debug("  ", e.expected_class.__name__, repr(e.src_path))
+
+        found_events = []
+        while True:
+            # Read all the available events until we timeout.
+            try:
+                event = self.event_queue.get(timeout=timeout)[0]
+            except Empty:
+                self._debug("event queue timeout")
+                break
+            self._debug("got:", event.__class__.__name__, repr(event.src_path))
+            found_events.append(event)
+
+        if self._validate_order:
+            self._check_events_with_order(found_events)
+        else:
+            self._check_events_without_order(found_events)
+
+    def __enter__(self) -> _EventsChecker:  # noqa: PYI034
+        # This could be annotated to return `Self` but that type is available
+        # only in Python 3.11+.
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        if exc_value is None:
+            # do check only if there is no error
+            self.check_events()
